### PR TITLE
drivers: syscon: Add support for multiple regions & emulator

### LINKED ISSUE #38737: drivers: syscon: missing implementation
**Describe the bug**
The syscon API (`include/drivers/syscon.h`) added in commit bc305984565a8f30e8afd02026dc99c16853a362 lacks an implementation. The generic syscon driver (`drivers/syscon/syscon.c`) implements the API functions, but uses different naming, so linking will fail.

This was never caught by CI since there are no in-tree users/testcases/samples using the syscon driver API.

The low-impact bug fix would just be renaming the functions in `drivers/syscon/syscon.c` to match those of the API, but a new approach, which addresses both the missing implementation, missing tests and generalizes the functionality even more, was already approved (#37424).

I suggest we merge #37424 for v2.7.0.

**Impact**
The syscon API lacks an in-tree implementation.

### PR DESCRIPTION
**Pull request by yperess**
1. Add support for multiple syscon entries (as a side effect, this also fixed syscon.c implementations which weren't being linked to their syscon.h counterparts).
2. Add a syscon emulator and unit tests.

Signed-off-by: Yuval Peress <peress@chromium.org>

### CODE REVIEW

**Comment by carlescufi** [comment]
Would this belong under `hwinfo` instead?

**Comment by yperess** [comment]
> Would this belong under `hwinfo` instead?

I don't think so, hwinfo seems to imply read only and specific to the hardware. This is a device register that the application is free to read/write to. It is generally used (in my experience) to contain application specific data instead of hardware descriptive data. @MulinChao I'm open to moving it if someone from Nuvoton wants to weigh in.

File: dts/arm/nuvoton/npcx.dtsi [hunk 1]
```diff
@@ -93,6 +93,12 @@
 			label = "SCFG";
 		};
 
+		mdc: mdc@4000c000 {
+			compatible = "nuvoton,npcx-mdc";
+			reg = <0x4000c000 0x4>;
```
Comment (WealianLiao) [inline dts/arm/nuvoton/npcx.dtsi:98]: Does the range cover the registers which you want to use?

File: drivers/mdc/mdc_npcx.c [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+static int mdc_npcx_read(const struct device *dev, size_t offset, size_t size, uint8_t *out)
+{
+	const struct mdc_npcx_config *config = (const struct mdc_npcx_config *)(dev->config);
+	volatile uint8_t *reg_addr = config->addr;
+
+	if (size < 1 || offset + size > config->size) {
+		return -EFAULT;
+	}
+
+	memcpy(out, reg_addr + offset, size);
```
Comment (WealianLiao) [inline drivers/mdc/mdc_npcx.c:34]: I don't recommend that read/write a register with different sizes access (e.g., read 32-bit data from an 8-bit register). Can we just add API for Bootloader scratch get & set. So we can access these registers with byte-length.

**Review by WealianLiao — COMMENTED** [review]
There is a MISC folder in the API folder. Is it a different meaning between MDC & MISC?

File: dts/arm/nuvoton/npcx.dtsi (hunk 1 above)
Comment (yperess) [reply dts/arm/nuvoton/npcx.dtsi:98]: I see, I misread the datasheet, the size here should be 16.

File: drivers/mdc/mdc_npcx.c (hunk 2 above)
Comment (yperess) [reply drivers/mdc/mdc_npcx.c:34]: Done

**Comment by yperess** [comment]
> There is a MISC folder in the API folder. Is it a different meaning between MDC & MISC?

I was tempted to put it in misc at first but I think it doesn't belong there. the `misc/` directory seems to hold simple drivers that don't really fit in anywhere. From what I can tell (hard to speculate since there's no README in that directory) it is really for drivers that are likely just used for one specific chip and will be unlikely to have multiple implementations. I thought that the MDC driver doesn't quite fit that, but I'm open to moving it.

**Comment by yperess** [comment]
Seems that the build error is unrelated.

**Comment by WealianLiao** [comment]
> > There is a MISC folder in the API folder. Is it a different meaning between MDC & MISC?
> 
> I was tempted to put it in misc at first but I think it doesn't belong there. the `misc/` directory seems to hold simple drivers that don't really fit in anywhere. From what I can tell (hard to speculate since there's no README in that directory) it is really for drivers that are likely just used for one specific chip and will be unlikely to have multiple implementations. I thought that the MDC driver doesn't quite fit that, but I'm open to moving it.

Recently, a 'syscon' driver was introduced. System controller node is a register region containing a set of miscellaneous registers. The 'syscon' API provides the regs access function. I think this is very close to the API in this PR. If you agree with that, how about change to use 'syscon'. 

For syscon, it seems the register only access with 32-bits. I think the API is under the Experimental Level. Maybe we can ask the owner whether we can add the register size as an API parameter.

**Comment by yperess** [comment]
> For syscon, it seems the register only access with 32-bits. I think the API is under the Experimental Level. Maybe we can ask the owner whether we can add the register size as an API parameter.

Yes, that does look close. I'll update this PR to update the API for `syscon` and see what the owner thinks.

**Comment by yperess** [comment]
@carlocaione it seems that we've crossed paths, I looked briefly at modifying the `syscon` driver to allow for miscellaneous register read/write such as provided in this PR but I can't say I fully understand the `MMIO` macros used there. What are your thoughts about:

1. Making the API more generic to be able to read/write _n_ bytes?
2. Adding a `get_size` API
3. Making the API more generic to allow multiple regions via `DT_INST_FOREACH_STATUS_OKAY`
4. I think we'll need this but here's where my understanding of the `MMIO` APIs fall short, but I think that in order to create an emulator for this we'll still need a different binding such as `syscon-emul`. I'm not seeing anything where we'd be able to specify the address/size of the memory region and the region is also created for the emulator.

**Comment by carlocaione** [comment]
> For syscon, it seems the register only access with 32-bits. I think the API is under the Experimental Level. Maybe we can ask the owner whether we can add the register size as an API parameter

It's not experimental level, it's just a bare minimum syscon driver without anything fancy on top (such as `regmap` & co) and really aimed to R/W a MMIO region. In this case you want to access the "registers" in the syscon region, so in the 99% of the cases you do not need to write anything bigger / smaller than 32-bits registers. For reference see the linux `syscon` driver, that is using the `regmap` driver to read / write 32-bits registers.

> Making the API more generic to be able to read/write n bytes?

In that case I don't think the `syscon` driver is the right tool for you. By definition (quoting from the Linux documentation) the `System controller node represents a register region containing a set of miscellaneous registers.`, it's not a driver used to read / write a generic amount of data from / to memory.

If you want instead to read / write several registers at once you can build that on top of the syscon driver simply calling the `write()` / `read()` functions several times or maybe writing a wrapper on top but, again, the syscon driver is aimed at reading / writing one register at time.

> Making the API more generic to allow multiple regions via DT_INST_FOREACH_STATUS_OKAY

Yeah, I don't remember why we assumed we only have one single `syscon` region, but feel free to extend it.

> I think we'll need this but here's where my understanding of the MMIO APIs fall short, but I think that in order to create an emulator for this we'll still need a different binding such as syscon-emul. I'm not seeing anything where we'd be able to specify the address/size of the memory region and the region is also created for the emulator.

I didn't get this. What do you mean with "emulator"?

**Review by carlocaione — CHANGES_REQUESTED** [review]
This needs more discussion

**Comment by yperess** [comment]
> > For syscon, it seems the register only access with 32-bits. I think the API is under the Experimental Level. Maybe we can ask the owner whether we can add the register size as an API parameter
> 
> It's not experimental level, it's just a bare minimum syscon driver without anything fancy on top (such as `regmap` & co) and really aimed to R/W a MMIO region. In this case you want to access the "registers" in the syscon region, so in the 99% of the cases you do not need to write anything bigger / smaller than 32-bits registers. For reference see the linux `syscon` driver, that is using the `regmap` driver to read / write 32-bits registers.
> 
> > Making the API more generic to be able to read/write n bytes?
> 
> In that case I don't think the `syscon` driver is the right tool for you. By definition (quoting from the Linux documentation) the `System controller node represents a register region containing a set of miscellaneous registers.`, it's not a driver used to read / write a generic amount of data from / to memory.
> 
> If you want instead to read / write several registers at once you can build that on top of the syscon driver simply calling the `write()` / `read()` functions several times or maybe writing a wrapper on top but, again, the syscon driver is aimed at reading / writing one register at time.
We can work with that, I don't see an issue there.
> 
> > Making the API more generic to allow multiple regions via DT_INST_FOREACH_STATUS_OKAY
> 
> Yeah, I don't remember why we assumed we only have one single `syscon` region, but feel free to extend it.
ACK
> 
> > I think we'll need this but here's where my understanding of the MMIO APIs fall short, but I think that in order to create an emulator for this we'll still need a different binding such as syscon-emul. I'm not seeing anything where we'd be able to specify the address/size of the memory region and the region is also created for the emulator.
> 
> I didn't get this. What do you mean with "emulator"?
When we write tests that depend on `syscon` API we'll need an entry in devicetree so that the code being tested will continue to work. You can see for example in this PR that the emulator basically just allocates _N_ bytes based on the region size. The API just reads/writes from there.

My understanding of MMIO API is that it only holds the pointer to the memory, but there's no version of the API that also allocates the memory such that host testing can function. By looking at what the `emulator` option does in this PR would you agree that we need a separate driver implementation for the emulator?

**Comment by carlocaione** [comment]
> When we write tests that depend on `syscon` API we'll need an entry in devicetree so that the code being tested will continue to work. You can see for example in this PR that the emulator basically just allocates _N_ bytes based on the region size. The API just reads/writes from there.

So, if I'm understanding this correctly you introduced a mockup driver basically a copy of another actual driver with the only intent for it to be used during the test on posix platform? I'll leave to @carlescufi @galak & co to decide whether this is a legit way forward or not.

> My understanding of MMIO API is that it only holds the pointer to the memory, but there's no version of the API that also allocates the memory such that host testing can function.

This is correct. On the other hand it would be weird to have a copy of each single driver writing to RAM instead of MMIO with the sole purpose of having a test passing on posix instead of real hardware. Makes no sense to me.

> By looking at what the `emulator` option does in this PR would you agree that we need a separate driver implementation for the emulator?

Again, it is odd to me this approach. Let's say that I have a piece of hardware and I write some driver for some peripheral. I would be pissed if I had to rewrite the same driver avoiding MMIO R/W and instead using regular memory accesses just to be able to pass a test on the native posix platform.

**Review by carlocaione — CHANGES_REQUESTED** [review]
Please, spilt this patch in two different PRs. The first only to modify the current syscon driver and another one to add the `emulator`.

File: drivers/syscon/CMakeLists.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
 # SPDX-License-Identifier: Apache-2.0
 
 zephyr_library()
-zephyr_library_sources_ifdef(CONFIG_SYSCON          syscon.c)
+zephyr_library_sources_ifdef(CONFIG_SYSCON_GENERIC          syscon.c)
```
Comment (carlocaione) [inline drivers/syscon/CMakeLists.txt:4]: Leave `CONFIG_SYSCON` for now. To be decided if there will be ever other types.

File: drivers/syscon/syscon.c [hunk 4]
```diff
@@ -12,106 +12,112 @@
...
+	*addr = DEVICE_MMIO_GET(dev);
 	return 0;
 }
 
-int syscon_generic_read_reg(uint16_t reg, uint32_t *val)
+static int syscon_generic_read_reg(const struct device *dev, uint16_t reg, uint32_t *val)
 {
 	struct syscon_generic_data *data;
+	const struct syscon_generic_config *config;
 	uintptr_t base_address;
```
Comment (carlocaione) [inline drivers/syscon/syscon.c:41]: reverse christmas tree please.

File: drivers/syscon/syscon.c [hunk 5]
```diff
@@ -12,106 +12,112 @@
...
 
-	if (sanitize_reg(&reg, data->size)) {
+	if (syscon_sanitize_reg(&reg, data->size, config->reg_width)) {
 		return -EINVAL;
 	}
 
-	base_address = DEVICE_MMIO_GET(syscon_dev);
+	base_address = DEVICE_MMIO_GET(dev);
 
 	*val = sys_read32(base_address + reg);
```
Comment (carlocaione) [inline drivers/syscon/syscon.c:60]: I don't get it. Shouldn't this be `sys_read{8,16,32}` according to the value of `config->reg_width`?

File: drivers/syscon/syscon.c [hunk 6]
```diff
@@ -12,106 +12,112 @@
...
 
 	return 0;
 }
 
-int syscon_generic_write_reg(uint16_t reg, uint32_t val)
+static int syscon_generic_write_reg(const struct device *dev, uint16_t reg, uint32_t val)
 {
 	struct syscon_generic_data *data;
+	const struct syscon_generic_config *config;
 	uintptr_t base_address;
```
Comment (carlocaione) [inline drivers/syscon/syscon.c:69]: ditto

File: drivers/syscon/syscon.c [hunk 7]
```diff
@@ -12,106 +12,112 @@
...
 
-	if (sanitize_reg(&reg, data->size)) {
+	if (syscon_sanitize_reg(&reg, data->size, config->reg_width)) {
 		return -EINVAL;
 	}
 
-	base_address = DEVICE_MMIO_GET(syscon_dev);
+	base_address = DEVICE_MMIO_GET(dev);
 
 	sys_write32(val, (base_address + reg));
```
Comment (carlocaione) [inline drivers/syscon/syscon.c:84]: ditto

File: dts/arm/nuvoton/npcx.dtsi [hunk 8]
```diff
@@ -101,6 +101,20 @@
 			label = "SCFG";
 		};
 
+		mdc: mdc@4000c000 {
+			compatible = "syscon";
+			reg = <0x4000c000 0xa>;
+			reg-io-width = <1>;
+			label = "MDC";
```
Comment (carlocaione) [inline dts/arm/nuvoton/npcx.dtsi:108]: Are the labels needed?

File: dts/bindings/syscon/syscon.yaml [hunk 9]
```diff
@@ -10,3 +10,8 @@ include: base.yaml
 properties:
     reg:
       required: true
+    reg-io-width:
+      type: int
+      default: 4
```
Comment (henrikbrixandersen) [inline dts/bindings/syscon/syscon.yaml:15]: The use of `default` is discouraged. The same behaviour can be obtained by using `DT_INST_PROP_OR()` in the C implementation file.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/syscon/syscon.c (hunk 4 above)
Comment (yperess) [reply drivers/syscon/syscon.c:41]: Done

File: drivers/syscon/syscon.c (hunk 5 above)
Comment (yperess) [reply drivers/syscon/syscon.c:60]: Done

File: drivers/syscon/syscon.c (hunk 6 above)
Comment (yperess) [reply drivers/syscon/syscon.c:69]: Done

File: drivers/syscon/syscon.c (hunk 7 above)
Comment (yperess) [reply drivers/syscon/syscon.c:84]: Done

File: dts/bindings/syscon/syscon.yaml (hunk 9 above)
Comment (yperess) [reply dts/bindings/syscon/syscon.yaml:15]: Done

File: dts/arm/nuvoton/npcx.dtsi (hunk 8 above)
Comment (yperess) [reply dts/arm/nuvoton/npcx.dtsi:108]: I'm not sure it's _needed_ but it does follow the pattern in the rest of the file so I think we shouldn't deviate.

File: drivers/syscon/CMakeLists.txt (hunk 3 above)
Comment (yperess) [reply drivers/syscon/CMakeLists.txt]: Done

File: dts/arm/nuvoton/npcx.dtsi (hunk 8 above)
Comment (carlocaione) [reply dts/arm/nuvoton/npcx.dtsi:108]: We are actually trying to move away from labels so remove it if you do not plan to use them.

Comment (yperess) [reply dts/arm/nuvoton/npcx.dtsi:108]: Done

File: dts/arm/nuvoton/npcx.dtsi [hunk 10]
```diff
@@ -101,6 +101,18 @@
...
+			reg = <0x4000c000 0xa>;
+			reg-io-width = <1>;
+		};
+
+		mdc_header: mdc@4000c00a {
+			compatible = "syscon";
+			reg = <0x4000c00a 0x4>;
+			reg-io-width = <2>;
+		};
+
```
Comment (carlocaione) [inline dts/arm/nuvoton/npcx.dtsi:115]: The DTS change should go in a different commit.

Comment (yperess) [reply dts/arm/nuvoton/npcx.dtsi:115]: Done

**Review by carlocaione — APPROVED** [review]

File: drivers/syscon/syscon.c [hunk 11]
```diff
@@ -12,106 +12,136 @@
...
+	static const struct syscon_generic_config syscon_generic_config_##inst = {                 \
+		DEVICE_MMIO_ROM_INIT(DT_DRV_INST(inst)),                                           \
+		.reg_width = DT_INST_PROP_OR(inst, reg_io_width, 4),                               \
+	};                                                                                         \
+	static struct syscon_generic_data syscon_generic_data_##inst = {                           \
+		.size = DT_REG_SIZE(DT_DRV_INST(inst)),                                            \
+	};                                                                                         \
+	DEVICE_DT_INST_DEFINE(inst, syscon_generic_init, NULL, &syscon_generic_data_##inst,        \
+			      &syscon_generic_config_##inst, PRE_KERNEL_1,                         \
+			      CONFIG_SYSCON_INIT_PRIORITY_DEVICE, syscon_generic_driver_api);
```
Comment (WealianLiao) [inline drivers/syscon/syscon.c:145]: &syscon_generic_driver_api

Comment (yperess) [reply drivers/syscon/syscon.c]: Thanks for the catch. I added a `build_only` test to make sure that the driver builds.

@carlocaione what are your thoughts about providing a more robust test for this? AFAICT what we would need is a section in the linker that we know the address of ahead of time. It seems that `__attribute__((at(???)))` doesn't work. Any ideas?

Comment (carlocaione) [reply drivers/syscon/syscon.c]: you could use a `reserved-memory` node + QEMU + `syscon` to create a test. Pretty much something like https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/boards/qemu_cortex_a53/reserved_memory but adding the `syscon` on top.

Comment (yperess) [reply drivers/syscon/syscon.c]: Odd, it doesn't seem to go at the address I'm specifying. I'm seeing:

```
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/syscon/src/main.c:20: test_base_address: (addr not equal to expected_addr)
addr=0x407ff000, exoected_addr=0x42000000
```

Is that expected? I was hoping to check that `syscon_get_base()` returns the address I set via:

```
/ {
	reserved-memory {
		compatible = "reserved-memory";
		#address-cells = <1>;
		#size-cells = <1>;
		status = "okay";
		ranges;

		res0: res@42000000 {
			reg = <0x42000000 0x8>;
			label = "res0";
		};
	};
	syscon: syscon@42000000 {
		compatible = "syscon";
		status = "okay";
		reg = <0x42000000 0x8>;
		reg-io-width = <1>;
	};
};
```

Comment (yperess) [reply drivers/syscon/syscon.c]: It seems that I'm also not able to write to this region. I tried:

```
uint8_t var_in_res0[DT_REG_SIZE(DT_NODELABEL(syscon))] __attribute((__section__(".res0")));

void test_main(void)
{
  printk("var_in_res0@0x%lx\n", (uintptr_t)var_in_res0);
  var_in_res0[0] = 0;
}
```

Which results in:

```
var_in_res0@0x42000000
E: ELR_ELn: 0x000000004000164c
E: ESR_ELn: 0x0000000096000046
E:   EC:  0x25 (Data Abort taken without a change in Exception level)
E:   IL:  0x1
E:   ISS: 0x46
E: FAR_ELn: 0x0000000042000000
E: TPIDRRO: 0x0100000040014990
E: x0:  0x0000000000000017  x1:  0x000000000000000a
E: x2:  0x000000004000a485  x3:  0x000000004005e599
E: x4:  0x000000000000000e  x5:  0x0000000000000002
E: x6:  0x0000000000000004  x7:  0x0000000000000000
E: x8:  0x0000000000000037  x9:  0x00000000ffffffff
E: x10: 0x0000000000000000  x11: 0x0000000000000000
E: x12: 0x0000000000000000  x13: 0x0000000000000000
E: x14: 0x0000000000000000  x15: 0x0000000000000000
E: x16: 0x0000000040005280  x17: 0x0000000000000000
E: x18: 0x0000000000000000  x30: 0x000000004000164c
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x40014640 (main)
E: Halting system
```

So this is telling me that `var_in_res` was actually placed at `0x42000000` as expected but `syscon_get_base()` is returning a different value. I also thought that this might be an artifact of `DEVICE_MMIO_IS_IN_RAM` not being set and the region being placed in ROM, but even with turning on `CONFIG_MMU=y` this is still failing.

Comment (carlocaione) [reply drivers/syscon/syscon.c]: The address returned by `syscon_get_base()` is expected to be different from `0x42000000`, because that's a virtual address while `0x42000000` is the physical address. You are defining a reserved region specifying the physical address but the syscon driver is io-mapping that region before returning to you the virtual address, the two address can be different. You cannot directly write to `0x42000000` without triggering a fault.

What you have to do is always using the address returned by `syscon_get_base()` and nothing else. You can write a certain amount of bytes at that address with direct memory access and then you should try to get back using the sysycon driver. Or the opposite: you use the syscon driver to write some data and retrieve that back using direct memory access. Bottom line: you never access directly the memory using the physical address.

This is of course valid if you have the MMU enabled (as you should if you are using A53).

Comment (yperess) [reply drivers/syscon/syscon.c]: Thanks for explaining. I've updated the test.

File: tests/drivers/syscon/src/main.c [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+	for (uint32_t i = 0; i < ARRAY_SIZE(var_in_res0); ++i) {
+		zassert_ok(syscon_write_reg(dev, i, i), NULL);
+		zassert_equal(((uint8_t *)base_addr)[i], i, NULL);
+	}
+}
+
+void test_main(void)
+{
+	ztest_test_suite(syscon, ztest_unit_test(test_size), ztest_unit_test(test_out_of_bounds),
+			 ztest_unit_test(test_read), ztest_unit_test(test_write));
```
Comment (WealianLiao) [inline tests/drivers/syscon/src/main.c:62]: One test per line

File: tests/drivers/syscon/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2021 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(bbram)
```
Comment (WealianLiao) [inline tests/drivers/syscon/CMakeLists.txt:7]: syscon

File: tests/drivers/syscon/src/main.c (hunk 12 above)
Comment (yperess) [reply tests/drivers/syscon/src/main.c:62]: Ugh, clang-format got me :)

File: tests/drivers/syscon/CMakeLists.txt (hunk 13 above)
Comment (yperess) [reply tests/drivers/syscon/CMakeLists.txt:7]: Done

**Review by WealianLiao — APPROVED** [review]
LGTM.

**Comment by yperess** [comment]
@nashif Can this merge? I know we're in the stabilization period, but:

* This is a fixes the syscon driver as prior the header function `syscon_get_base` wasn't implemented, but instead had an implementation as `syscon_generic_get_base`.
* It also didn't support having more than one entry in the device tree. So creating multiple uses of syscon would have undefined behavior (assuming the above was fixed).
* It adds tests.

**Review by jackrosenthal — APPROVED** [review]

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks really, thanks. A few minor suggestions:

**Comment by yperess** [comment]
> Looks really, thanks. A few minor suggestions:

Thanks, applied the suggestions.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by jackrosenthal — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]
Nice work!
